README.md
# Google Photos Sync/Backup

This Go program downloads a (filtered) collection of a Google Photos account into a local directory
hierarchy.

Unfortunately, Google does not offer any simple solution to just download photos by certain filters,
so this tool fills this gap.

Luckily, Google Photos comes with a REST API, which allows querying and downloading the Photos library.
This script is a very simple approach to fetch your Google Photos automatically.

**Note**: This is a very early version. It is NOT tested thoroughly, and it is NOT implemented in a secure and stable way! Be aware of daemons!

- [Requirements](#requirements)
- [Setup](#setup)
  - [Step 1: Enable the Photos API](#step-1-enable-the-photos-api)
  - [Step 2: Create a Google OAuth 2.0 Client ID](#step-2-create-a-google-oauth-20-client-id)
  - [Step 3: Build the program](#step-3-build-the-program)
- [Usage](#usage)
  - [Command line flags](#command-line-flags)
- [Limitations](#limitations)
- [More on building](#more-on-building)
  - [Releases](#releases)
- [Notes](#notes)
- [Wishlist](#wishlist)


## Requirements

* The Go tools, if you want to compile it by yourself
* A Google Photos account
* A Google Photos account and access to the Google Developer Console (https://console.developers.google.com/)

## Setup

### Step 1: Enable the Photos API

You need to enable the Photos API in order to use it.

1. Log in to https://console.developers.google.com/
2. In the Dashboard, click on "Enable APIs and Services":<br/>
   ![Enable APIs Menu](./doc/img/gcloud_enable_apis.png)
3. Search for "Photos", and select the Photos entry:<br />
   ![Photos Menu Entry](./doc/img/gcloud_photos_entry.png)
4. Click the "Enable API" button

Return to the Developer Console Dashboard.

### Step 2: Create a Google OAuth 2.0 Client ID

The Google Photos API only works with OAuth 2.0. You need to create an OAuth 2.0 Client ID and download your secret credentials
in order to use them with the pyhton tool.

1. In the Developer Console, click "Credentials" in the left menu, then select "Create Credentials" &gt; "OAuth Client ID":<br/>
   ![OAuth Client ID Menu Entry](./doc/img/google-cloud-api-create_login_menu.png)
2. Select "Other" in the Create OAuth client ID menu, and provide a name. This name is not important for the application, just for you to
   recognize it later.<br/>
   ![OAuth Client ID creation](./doc/img/google-cloud-api-create-oauth-client-id.png)
3. Click "Create".
4. Back in the Dashboard's "Credentials" menu, click the download button on the newly created client it. The JSON file you get contains
   the secret client tokens. KEEP IT IN A SAFE PLACE, it contains the keys that allow access the Google Photos library! <br />
   ![OAuth Client ID credential download](./doc/img/google-cloud-api-download-client_creds.png)


### Step 3: Build the program

```bash
$ git clone git@github.com:bylexus/gphotos-sync.git
$ cd gphotos-sync
$ go build
```

## Usage

Once the setup is complete, you start the tool with the following command:

```bash
$ ./gphotos-sync /path/to/photo/backup/folder
```

If you start the tool for the first time, it initiates the Google Authentication workflow:

1. You will be asked for your Google Client ID and Google Client Secret. Enther those values,
   which you can find in the Developer Console <https://console.cloud.google.com>
2. Your Browser will start and asks you for your Google permissions for the gphotos-sync app
3. After the authentication was successful, you can close the browser and the command line
   tool should resume its operation.

***NOTE***

The Google Authentication information is stored locally in a settings file in the User's config directory (`secrets.json`). 
This file contains sensitive information (your authentication token). Do NOT share this file!

If all goes well, the tool begins to transfer your photos. Enjoy!

### Command line flags

The following command line flags are interpreted to configure the sync:

**`-h, --help`**:

Prints a brief help

**`--date=YYYY[-MM[-DD]]`**:

Filter by date. The date is either a full date representing a specific day, or a part of it: So `2023` will sync the whole year, while `2023-04` will sync only April, 2023.
The `--date=` parameter can be defined up to 5 times (google api limitation).

**`--date-range=YYYY[-MM[-DD]]:YYYY[-MM[-DD]]`**:

Filter by a date range. Define start and end date in the format YYYY-MM-DD:YYYY-MM-DD. Partial dates possible, BUT startDate and endDates must be of the same format.

Example: 

`gphotos-sync --date-range=2023-04-01:2023-05-15 dest/` will fetch all photos from April 1st, 2023, to Mid may, 2023
The `--date-range=` parameter can be defined up to 5 times (google api limitation).

**`-t nr, --threads=nr`**:

Number of concurrent download threads to use, defaults to 5. Use a smaller number on slow network connections.
If you have a 100mbps connection or more, 5 or more threads are much faster.

**`-f, --force`**:

Force file override: If a local file already exists with the same name, it is overridden. Default is to skip.

**`-n, --force-newer`**:

Force file override for newer remote files only: If a local file already exists with the same name, it is overridden
if the remote file is newer (and only then). Default is to skip.

## Limitations

* Downloaded photos include the original EXIF data, EXCEPT the location data: This seems to be a Google Photos API limitation, officially enforced by Google.
  To this moment, there seems to be no solution around that.

## More on building

You can use the Makefile to help building the tool and cross-compile to several os'es and architectures:

```sh
# default build:
make

# build for windows:
make gphotos-sync.exe

# create all releases:
make release

# clean build artifacts:
make clean
```

Note that the Makefile is only supported on Unix-like operating systems.

### Releases



## Notes

* This tool is in an early experimental stage. Do NOT expect it to work flawlessly.
* I have not tested it over a longer period of time (e.g. until the client tokens expire). I do not yet know what happens then.
* I will implement some additional config params in the future, see Wishlist below

## Wishlist

See [CHANGELOG](./CHANGELOG.md)


If you have any more ideas what this tool should be able to do, please drop me an issue or a note.

(c) 2023 <alex-gsync@alexi.ch>

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Wishlist]

- credentials store encryption
- output folder configuration: folder level generation by template (e.g. "{year}/{month}")
- Filters: by Media Type
- Extract photo metadata from google api and embed them as EXIF data
- Extract GPS data: Unfortunately, it seems that those are NOT available through the Photos API. It _may_ be a possibility to
  use Google Drive: https://www.labnol.org/code/20059-image-exif-and-location
- Improve logging and error handling: Separate errors on stderr, use a logger
- Sync back: Upload changed / missing photos to Google Photos
- date filters: remove limitation of max 5 date filters: create multiple jobs when needed

## [0.2.0] - 2023-06-20

### Fixed

- Fixing Date bug: 1st and last of day/month caused a wrong format error

### Added

- Filter by date range: It is now possible to filter by date range: `--date-range=2023-04-01:2023-05-15`
- cross-compile releases: use `make release` to build multiple binaries for windows, linux, macos

## [0.1.0] - 2023-06-17

This is the first release ever - the [Added] section below just lists the actual features.

### Added

- downloads Google Photo media items to a local folder
- creates a folder for each year
- supported files: Photos and Videos, as supported by Google Photos
- support for multiple download threads
- support for skip / override existing / newer files
- stores the google api tokens in a local credentials store


lib/app_config.go
package lib

import (
	"fmt"
	"os"
)

type CmdOptions struct {
	Date               []string `long:"date" value-name:"YYYY[-MM[-DD]]" description:"Set a single date filter, in the format YYYY-MM-DD. Partial date possible.\nExample: --date=2023 will fetch all photos from the year 2023."`
	DateRange          []string `long:"date-range" value-name:"YYYY[-MM[-DD]]:YYYY[-MM[-DD]]" description:"Filter by a date range. Define start and end date in the format YYYY-MM-DD:YYYY-MM-DD. Partial dates possible, BUT startDate and endDates must be of the same format.\nExample: --date-range=2023-04-01:2023-05-15 will fetch all photos from April, 2023, to Mid may, 2023"`
	ForceOverride      bool     `long:"force" short:"f" description:"Overrides local files in any case. Default is skip if a file exists locally."`
	ForceNewerOverride bool     `long:"force-newer" short:"n" description:"Override local files only if the remote file is newer, skip otherwise"`
	NrOfThreads        int      `long:"threads" short:"t" default:"5" value-name:"nr" description:"Number of download threads to use"`
}

type AppConfig struct {
	BaseOutputPath string
	Secrets        *Secrets

	DateFilter         []string
	DateRangeFilter    []string
	ForceOverride      bool
	ForceNewerOverride bool
	NrOfThreads        int
}

func CreateAppConfig(args []string, opts CmdOptions) AppConfig {
	if len(args) < 1 {
		fmt.Fprintf(os.Stderr, "Output path missing.\n")
		os.Exit(1)
	}
	outPath := args[0]
	err := os.MkdirAll(outPath, 0755)
	if err != nil {
		fmt.Fprintf(os.Stderr, "%v\n", err)
		os.Exit(1)
	}
	conf := AppConfig{
		BaseOutputPath:     outPath,
		Secrets:            LoadSecrets(),
		DateFilter:         opts.Date,
		DateRangeFilter:    opts.DateRange,
		ForceOverride:      opts.ForceOverride,
		ForceNewerOverride: opts.ForceNewerOverride,
		NrOfThreads:        opts.NrOfThreads,
	}
	if conf.NrOfThreads < 1 {
		conf.NrOfThreads = 1
	}
	if err := conf.Secrets.EnsureUserSecrets(); err != nil {
		panic(err)
	}

	return conf
}
